enumerables.rb

```diff
@@ -1,37 +1,72 @@
 module Enumerables
+
   def my_each
-    count = 0
-    while count < self.size do
-      yield(self[count])
-      count += 1
+    return to_enum unless block_given?
+
+    for i in 0..self.length
+      yield(self[i])
     end
-    self
   end
 
   def my_each_with_index
-    
+    return to_enum unless block_given?
+    my_index = 0
+    self.my_each do |i|
+      yield i, my_index
+      my_index += 1
+    end
   end
```

Comment:
- [ ] enumerables #my_each_with_index does not mutate the original array
     Failure/Error: array.my_each_with_index { |num| num + 1 }

- [ ]  enumerables #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self
     Failure/Error: let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }

- [ ] enumerables #my_each_with_index when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)
     
       expected: "Num: 4, idx: 0\nNum: 7, idx: 1\nNum: 3, idx: 2\nNum: 0, idx: 3\nNum: 2, idx: 4\nNum: 7, idx: 5\nNum:...: 5, idx: 94\nNum: 0, idx: 95\nNum: 5, idx: 96\nNum: 3, idx: 97\nNum: 2, idx: 98\nNum: 8, idx: 99\n"
            got: "Num: 4, idx: 0\nNum: 7, idx: 1\nNum: 3, idx: 2\nNum: 0, idx: 3\nNum: 2, idx: 4\nNum: 7, idx: 5\nNum:...: 0, idx: 95\nNum: 5, idx: 96\nNum: 3, idx: 97\nNum: 2, idx: 98\nNum: 8, idx: 99\nNum: , idx: 100\n"
     

- [ ] enumerables #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self
     Failure/Error: expect(range.my_each_with_index(&block)).to eq(range.each_with_index(&block))

- [ ]  enumerables #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
      Failure/Error: enum.my_each_with_index(&block)

- [ ] enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
      Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)
      
        expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
             got: "Num: , idx: 0\nNum: , idx: 1\nNum: , idx: 2\nNum: , idx: 3\nNum: , idx: 4\nNum: , idx: 5\n"